crypto/rsa/rsa_gen.c

```diff
@@ -462,9 +462,6 @@ static int rsa_keygen_pairwise_test(RSA *rsa, OSSL_CALLBACK *cb, void *cbarg)
                                        padding);
     if (ciphertxt_len <= 0)
         goto err;
-    if (ciphertxt_len == plaintxt_len
```

Comment:
I think the actual bug needs to be fixed for this - not just delete the code.
i.e- The ciphertext must be tested to be not equal to the plaintext.

Reply:
I'm pretty sure this is a copy paste in the wrong place problem.  The same block of code appears three lines further down after `decoded` has been initialised.

It's coverity 1462561 Uninitialized scalar variable:
```C
unsigned char *ciphertxt = NULL;
443    const unsigned char plaintxt[16] = {0};
   1. var_decl: Declaring variable decoded without initializer.
444    unsigned char decoded[256];
445    unsigned int decoded_len;
446    unsigned int plaintxt_len = (unsigned int)sizeof(plaintxt_len);
447    int padding = RSA_PKCS1_PADDING;
448    OSSL_SELF_TEST *st = NULL;
449
450    st = OSSL_SELF_TEST_new(cb, cbarg);
   2. Condition st == NULL, taking false branch.
451    if (st == NULL)
452        goto err;
453    OSSL_SELF_TEST_onbegin(st, OSSL_SELF_TEST_TYPE_PCT,
454                           OSSL_SELF_TEST_DESC_PCT_RSA_PKCS1);
455
456    ciphertxt_len = RSA_size(rsa);
457    ciphertxt = OPENSSL_zalloc(ciphertxt_len);
   3. Condition ciphertxt == NULL, taking false branch.
458    if (ciphertxt == NULL)
459        goto err;
460
461    ciphertxt_len = RSA_public_encrypt(plaintxt_len, plaintxt, ciphertxt, rsa,
462                                       padding);
   4. Condition ciphertxt_len <= 0, taking false branch.
463    if (ciphertxt_len <= 0)
464        goto err;
   5. Condition ciphertxt_len == plaintxt_len, taking true branch.
   CID 1462561 (#1 of 1): Uninitialized scalar variable (UNINIT)6. uninit_use_in_call: Using uninitialized element of array decoded when calling memcmp.
465    if (ciphertxt_len == plaintxt_len
466        && memcmp(decoded, plaintxt, plaintxt_len) == 0)
467        goto err;
468
469    OSSL_SELF_TEST_oncorrupt_byte(st, ciphertxt);
470
471    decoded_len = RSA_private_decrypt(ciphertxt_len, ciphertxt, decoded, rsa,
472                                      padding);
473    if (decoded_len != plaintxt_len
474        || memcmp(decoded, plaintxt,  decoded_len) != 0)
475        goto err;
476
477    ret = 1;
478 err:
479    OSSL_SELF_TEST_onend(st, ret);
480    OSSL_SELF_TEST_free(st);
481    OPENSSL_free(ciphertxt);
482
483    return ret;
```

Reply:
The intention for the last bit of code is to make sure the decrypted ciphertext matches the original input plaintext.
The intention of the middle check is that the encrypted cipher text does not match the plaintext i.e - the encryption actually does something. (its one of the check it says it should be doing).

Reply:
I guess it could be checking that the ciphertext is different to the plaintext???

Reply:
Put in a check that the ciphertxt is different to the plaintxt instead.